File: benchmarks/solutions/kotlin/s1-channel-pipeline.kt
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.*

fun CoroutineScope.producer() = produce {
    for (i in 1..20) send(i)
}

fun CoroutineScope.filterOdd(input: ReceiveChannel<Int>) = produce {
    for (v in input) {
        if (v % 2 == 1) send(v)
    }
}

fun main() = runBlocking {
    val numbers = producer()
    val odds = filterOdd(numbers)
    for (v in odds) {
        println(v)
    }
}

File: benchmarks/solutions/typescript/s1-channel-pipeline.ts
async function* producer(): AsyncGenerator<number> {
  for (let i = 1; i <= 20; i++) yield i;
}

async function* filterOdd(source: AsyncGenerator<number>): AsyncGenerator<number> {
  for await (const v of source) {
    if (v % 2 === 1) yield v;
  }
}

for await (const v of filterOdd(producer())) {
  console.log(v);
}
